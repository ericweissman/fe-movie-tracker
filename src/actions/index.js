export const getMovies = (movies) => ({
  type: 'GET_MOVIES',
  movies
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
})

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
})

export const populateFavorites = (favorites) => ({
  type: 'POPULATE_FAVORITES',
  favorites
})
