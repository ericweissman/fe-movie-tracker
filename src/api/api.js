export const fetchData = async (url) => {
  const response = await fetch(url);
  if(response.ok) {
    return response.json()
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}

export const postData = async (urlSuffix, data) => {
  const url = 'http://localhost:3000/api/users' + urlSuffix;
  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json'
    }
  })
  if(response.ok) {
    let data = response.json();
    return data
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}