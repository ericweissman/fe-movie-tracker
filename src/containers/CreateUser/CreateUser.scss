@font-face {
  font-family: 'Cereal';
  src: url(./AirbnbCereal-Black.ttf)
}

.createUser-form {
  color: $text;
  background-color: $accent;
  align-items: center;
  border: 2px solid rgb(242, 242, 242);
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  margin: 100px auto;
  width: 400px;

  h2 {
    font-family: 'Cereal'
  }

  p {
  font-family: 'Roboto';
  color: $primary;
  text-transform: uppercase;
  }

  input {
    border: 2px solid rgb(240, 240, 240);
    border-radius: 10px;
    height: 30px;
    margin-bottom: 15px;
    padding-left: 5px;
    width: 200px;
    outline: none;
  }

  button {
    border: none;
    border-radius: 10px;
    cursor: pointer;
    font-family: 'Roboto', sans-serif;
    font-size: 1rem;
    height: 30px;
    margin-bottom: 15px;
    transition-duration: .15s;
    width: 100px;

    &:hover {
      background-color: $text;
      color: $primary;    
      }
  }

  .link {
    font-family: 'Roboto', sans-serif;
    margin-bottom: 15px;
  }
}

@media only screen and (max-width: 449px) {
  .createUser-form {
    width: 285px
  }
}