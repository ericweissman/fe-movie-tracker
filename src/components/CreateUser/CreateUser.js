import React, { Component } from "react";
import { postData } from "../../api/api";
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux'
import { loginUser, populateFavorites } from '../../actions';

class CreateUser extends Component {
  constructor() {
    super();
    this.state = {
      name: "",
      email: "",
      password: "",
      passwordCheck: "",
      status: ""
    };
  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({
      [name]: value
    });
  };

  submitForm = async event => {
    const { name, email, password, passwordCheck } = this.state;
    event.preventDefault();
    try {
      if (password === passwordCheck) {
        var response = await postData("/new", { name, email, password });
        const currentUser = await postData('', { email, password })
        this.props.loginUser(currentUser.data)
        this.setState({ status: response.status })
        // this.props.loginUser()
      }
    } catch {
        this.setState({ status: "error"})
    }
  };

  render() {
    const { name, email, password, passwordCheck, status } = this.state;
    return (
      <form onSubmit={this.submitForm}>
        <h2>Create account</h2>
        <input onChange={this.handleChange} name="name" value={name} placeholder='name'/>
        <input onChange={this.handleChange} name="email" value={email} placeholder='email'/>
        <input onChange={this.handleChange} type='password' name="password" value={password} placeholder='password'/>
        <input onChange={this.handleChange} type='password' name="passwordCheck" value={passwordCheck} placeholder='confirm password'/>
        {password !== passwordCheck && <p>Passwords do not match</p> }
        {status === 'error' && <p>Email is taken</p>}
        {status === 'success' && <Redirect to='/' />}
        <button>SUBMIT</button>
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  loginUser: (user) => dispatch(loginUser(user)),
  populateFavorites: (favorites) => dispatch(populateFavorites(favorites))
})

export default connect(null, mapDispatchToProps)(CreateUser)