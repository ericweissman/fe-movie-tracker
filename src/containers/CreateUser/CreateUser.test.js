import React from 'react';
import { shallow } from 'enzyme';
import {CreateUser, mapDispatchToProps} from './CreateUser';
import { loginUser, populateFavorites } from '../../actions';
import * as api from '../../api/api'


describe('CreateUser', () => {
  let wrapper;
  const mockData = { data: {id:2, name: 'Eric', email: 'eric@aol.com'}, status: 'success'};
  api.postData = jest.fn(() => mockData);

  beforeEach(() => {
    wrapper = shallow(<CreateUser />);
    wrapper.setProps({loginUser: jest.fn(() => 'success')})
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have the proper default state', () => {
    const defaultState = {
      name: "",
      email: "",
      password: "",
      passwordCheck: "",
      status: ""
    }
    expect(wrapper.state()).toEqual(defaultState)
  });

  it('should update state when handleChange is called', () => {
    const mockEvent = {target: {name: 'name', value: 'eric'}}
    
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state().name).toEqual(mockEvent.target.value)
  });


  it('should set call postData with the correct params when creating a user', async() => {
    const expectedState = {
      name: 'Eric',
      email: 'e@aol.com',
      password: '2',
      passwordCheck: '2',
      status: 'success'
    }
    const param1 = '/new'
    const param2 = { name: 'Eric', email:'e@aol.com', password: '2'}
    const mockEvent = {preventDefault: jest.fn()};

    wrapper.setState(expectedState)
    wrapper.find('form').simulate('submit', mockEvent);
    await expect(api.postData).toHaveBeenCalledWith(param1, param2)
  })

  it.skip('should call loginUser with the correct params', async () => {
    //can't get to this level - only reads the first await postData
    const mockUser = {name: mockData.data.name, id: mockData.data.id, favorites: []};
    const mockEvent = { preventDefault: jest.fn() };
    wrapper.find('form').simulate('submit', mockEvent);
    await expect(api.postData).toHaveBeenCalledWith(mockUser)
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using a function from mapDispatchToProps', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = loginUser({name: 'Eric', email: 'eric@gmail.com'});

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.loginUser({ name: 'Eric', email: 'eric@gmail.com'})

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})