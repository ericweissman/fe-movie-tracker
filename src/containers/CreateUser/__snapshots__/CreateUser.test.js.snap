// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateUser should match the correct snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Create account
  </h2>
  <input
    name="name"
    onChange={[Function]}
    placeholder="name"
    value=""
  />
  <input
    name="email"
    onChange={[Function]}
    placeholder="email"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="password"
    type="password"
    value=""
  />
  <input
    name="passwordCheck"
    onChange={[Function]}
    placeholder="confirm password"
    type="password"
    value=""
  />
  <button>
    SUBMIT
  </button>
</form>
`;
